lib/zipkin-tracer/trace.rb

```diff
@@ -1,19 +1,132 @@
-require 'finagle-thrift/trace'
 require 'zipkin-tracer/zipkin_tracer_base'
 # Module with a mix of functions and overwrites from the finagle implementation:
 # https://github.com/twitter/finagle/blob/finagle-6.39.0/finagle-thrift/src/main/ruby/lib/finagle-thrift/trace.rb
 module Trace
+  # These methods and attr_accessor below are used as global configuration of this gem
+  # Most of these are set by the config class and then used around.
+  # TODO: Move this out of the Trace module , take out that extend self and be happier
+  extend self
   attr_accessor :trace_id_128bit
 
-  # We need this to access the tracer from the Faraday middleware.
   def self.tracer
     @tracer
   end
 
-  def sample_rate
+  def self.sample_rate
     @sample_rate
   end
 
+  def self.tracer=(tracer)
+    @tracer = tracer
+  end
+
+  def self.sample_rate=(sample_rate)
+    if sample_rate > 1 || sample_rate < 0
+      raise ArgumentError.new("sample rate must be [0,1]")
+    end
+    @sample_rate = sample_rate
+  end
+
+  def default_endpoint=(endpoint)
+    @default_endpoint = endpoint
+  end
+
+  def default_endpoint
+    @default_endpoint
+  end
+
+  # These classes all come from Finagle-thrift + some needed modifications (.to_h)
+  # Moved here as a first step, eventually move them out of the Trace module
+
+  class Annotation
+    CLIENT_SEND = "cs"
+    CLIENT_RECV = "cr"
+    SERVER_SEND = "ss"
+    SERVER_RECV = "sr"
+
+    attr_reader :value, :host, :timestamp
+    def initialize(value, host)
+      @timestamp = (Time.now.to_f * 1000 * 1000).to_i # micros
+      @value = value
+      @host = host
+    end
+
+    def to_h
+      {
+        value: @value,
+        timestamp: @timestamp,
+        endpoint: host.to_h
+      }
+    end
+  end
+
+  class BinaryAnnotation
+    SERVER_ADDRESS = 'sa'.freeze
+    URI = 'http.uri'.freeze
```

Comment:
how about switching to `http.url`?
https://zipkin.io/public/thrift/v1/zipkinCore.html#Const_HTTP_URL